packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1374 @@
+            // - `customData` can be anything
+            latestPartial[key] = partial[key];
+          } else {
+            latestPartial[key] = element[key];
+          }
+        }
+
+        return latestPartial;
+      };
+
+    const applyLatestChangesInternal = (
+      deltas: Map<string, Delta<ElementPartial>>,
+    ) => {
+      const modifiedDeltas = new Map<string, Delta<ElementPartial>>();
+
+      for (const [id, delta] of deltas.entries()) {
+        const existingElement = elements.get(id);
+
+        if (existingElement) {
+          const modifiedDelta = Delta.create(
+            delta.deleted,
+            delta.inserted,
+            modifier(existingElement),
+            modifierOptions,
+          );
+
+          modifiedDeltas.set(id, modifiedDelta);
+        } else {
+          // Keep whatever we had
+          modifiedDeltas.set(id, delta);
+        }
+      }
+
+      return modifiedDeltas;
+    };
+
+    const updated = applyLatestChangesInternal(this.updated);
+
+    return ElementsChange.create(this.added, this.removed, updated);
+  }
+
+  public applyTo(
+    elements: SceneElementsMap,
+    snapshot: Map<string, OrderedExcalidrawElement>,
+  ): [SceneElementsMap, boolean] {
+    let nextElements = toBrandedType<SceneElementsMap>(new Map(elements));
+
+    const changed = new Map<string, OrderedExcalidrawElement>();
+    const added = new Map<string, OrderedExcalidrawElement>();
+    const removed = new Map<string, OrderedExcalidrawElement>();
+
+    const flags = {
+      containsVisibleDifference: false,
+      containsZindexDifference: false,
+    };
+
+    const applyDelta = ElementsChange.createApplier(nextElements, flags);
+    const getElement = ElementsChange.createGetter(
+      nextElements,
+      snapshot,
+      flags,
+    );
+
+    for (const [id, delta] of this.removed.entries()) {
+      const existingElement = getElement(id, delta.inserted);
+
+      if (existingElement) {
+        const removedElement = applyDelta(existingElement, delta);
+        changed.set(removedElement.id, removedElement);
+        removed.set(removedElement.id, removedElement);
+      }
+    }
+
+    for (const [id, delta] of this.updated.entries()) {
+      const existingElement = getElement(id, delta.inserted);
+
+      if (existingElement) {
+        const updatedElement = applyDelta(existingElement, delta);
+        changed.set(updatedElement.id, updatedElement);
+      }
+    }
+
+    for (const [id, delta] of this.added.entries()) {
+      const existingElement = getElement(id, delta.inserted);
+
+      if (existingElement) {
+        const addedElement = applyDelta(existingElement, delta);
+        changed.set(addedElement.id, addedElement);
+        added.set(addedElement.id, addedElement);
+      }
+    }
+
+    // Performs mutations, but mostly on new instances of elements
+    // trying to avoid potential mutations of previous state leading to unexpected incosistencies
```

Comment:
Could we fallback to previous state if we detect inconsistencies (assuming we could somehow)?

Reply:
For now we will rely on restore, in the future we could persist the history stack so one could always undo/redo (even after refresh).